### ISSUE TITLE
inconsistent behavior between delegate_to: "" and delegate_to: "{{ undef() | default('') }}"

### ISSUE THREAD

**Issue by craph** [ansible/ansible#84332]
### Summary

Hello,

I'm opening this issue because I'm seeing a weird behavior of the `delegate_to` when it's combined with conditional set to `false`

For example, with conditional set to false, if delegate_to contains a variable that contains an empty string, I have the following error message : 
```
fatal: [localhost -> {{ my_var }}]: FAILED! => {"msg": "Empty hostname produced from delegate_to: \"{{ my_var  }}\""}
```

But instead, if delegate_to is defined like this `delegate_to: ""` there is no error message like the previous one. 

Why ???

Delegate_to shouldn't be evaluated when he's surrounded by conditional defined to false.

Knowing my playbook, I have some block of tasks with conditional. This conditional can be set to false for specific use case and then the variable inside the block shouldn't have to be set because we know that the block will be skipped.

### Issue Type

Bug Report

### Component Name

lib/ansible/vars/manager.py

### Ansible Version

```console
$ ansible --version
ansible [core 2.18.0]
  config file = None
  configured module search path = ['/home/ubuntu/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/ubuntu/.local/share/pipx/venvs/ansible-core/lib/python3.12/site-packages/ansible
  ansible collection location = /home/ubuntu/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/ubuntu/.local/bin/ansible
  python version = 3.12.3 (main, Sep 11 2024, 14:17:37) [GCC 13.2.0] (/home/ubuntu/.local/share/pipx/venvs/ansible-core/bin/python)
  jinja version = 3.1.4
  libyaml = True
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
```

### OS / Environment

Ubuntu, RHEL9

### Steps to Reproduce

Here is the block that produce the error
```yaml (paste below)
---
- name: Debug
  hosts: localhost
  vars:
    my_var: ""

 - name : Debug block
    when: false
    block:
      - name: Echo toto
        ansible.builtin.debug:
          msg: "toto should be skipped"
        delegate_to: "{{ my_var }}"

      - name: Echo titi
        ansible.builtin.debug:
          msg: "titi should be skipped"
        delegate_to: "{{ my_var }}"
```

Here is the block that doesn't produce the error
```yaml (paste below)
---
- name: Debug
  hosts: localhost

 - name : Debug block
    when: false
    block:
      - name: Echo toto
        ansible.builtin.debug:
          msg: "toto should be skipped"
        delegate_to: ""

      - name: Echo titi
        ansible.builtin.debug:
          msg: "titi should be skipped"
        delegate_to: ""
```

### Expected Results

Here is the expect result for the block that have delegate_to with variable :
```
TASK [Echo toto] *******************************************************************************************
skipping: [localhost]

TASK [Echo titi] *******************************************************************************************
skipping: [localhost]
```

### Actual Results

```console
Here is the result for the block that produce the error :

TASK [Echo toto] *******************************************************************************************
fatal: [localhost -> {{ my_var }}]: FAILED! => {"msg": "Empty hostname produced from delegate_to: \"{{ my_var }}\""}
```

Here is the result for the block that doesn't produce the error :
```
TASK [Echo toto] *******************************************************************************************
skipping: [localhost]

TASK [Echo titi] *******************************************************************************************
skipping: [localhost]
```
```


### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ansibot** [comment]
Files identified in the description:

None

If these files are incorrect, please update the `component name` section of the description or use the [component](https://github.com/ansible/ansibotmini#commands) bot command.

**Comment by craph** [comment]
@rayamos Same problem for me and also have the issue in version `2.15.9`

**Comment by craph** [comment]
Here is another issue related to the same behavior #83253 
@sivel why this is an expected behavior ?
In my example tests cases the conditional when is ignored when this is a variable inside the delegate_to. The when conditional mustn't be ignored.
Thank you very much for your help on this subject.
Best regards,

**Comment by mat-m** [comment]
I have a similar use case

```yaml
- name: Migration
  when:
    - do_migration | bool
  tags: synchronize_files
  block:
    - name: "Create ssh keygen temporarily on {{ old_server }}"
      community.crypto.openssh_keypair:
        path: "~{{ user }}/.ssh/id_ssh_4_mig"
      become: true
      become_user: "{{ user }}"
      delegate_to: "{{ old_server }}"
      tags: sshkey

```

When we do not process data migration, we do not have the `old_server` variable set, obviously. I totally agree to break if `do_migration` is `true`, but it should be transparent if it is `false`.
I didn't test, but I guess the task does not fail if `user` is not set for `become_user`.

**Comment by craph** [comment]
> I have a similar use case
> 
> ```yaml
> - name: Migration
>   when:
>     - do_migration | bool
>   tags: synchronize_files
>   block:
>     - name: "Create ssh keygen temporarily on {{ old_server }}"
>       community.crypto.openssh_keypair:
>         path: "~{{ user }}/.ssh/id_ssh_4_mig"
>       become: true
>       become_user: "{{ user }}"
>       delegate_to: "{{ old_server }}"
>       tags: sshkey
> ```
> 
> When we do not process data migration, we do not have the `old_server` variable set, obviously. I totally agree to break if `do_migration` is `true`, but it should be transparent if it is `false`. I didn't test, but I guess the task does not fail if `user` is not set for `become_user`.

Hi @mat-m ,
I have tested your sample and as expected, the task doesn't fail if I have `user: ""` the task is correctly skipped. 
But if delegate_to has the variable `old_server` set to `""` the task fails evenif `do_migration` is set to `false`

**Comment by uhmfa** [comment]
I am also experiencing this. It would be great to have a fix to get the expected behavior.
Thanks

**Comment by sivel** [comment]
After triaging this issue, we've agreed that behavior is inconsistent.  The plan forward is to also make `delegate_to: ""` an error to match the templated outcome.

The comment from https://github.com/ansible/ansible/issues/83253#issuecomment-2112764139 explains how to do this from a templating perspective.

From a non-templating perspective, I can see no reason why anyone would explicitly provide an empty string to `delegate_to`, so the workaround is simply to not do that.

**Comment by mat-m** [comment]
> After triaging this issue, we've agreed that behavior is inconsistent. The plan forward is to also make `delegate_to: ""` an error to match the templated outcome.
> 
> The comment from [#83253 (comment)](https://github.com/ansible/ansible/issues/83253#issuecomment-2112764139) explains how to do this from a templating perspective.
> 
> From a non-templating perspective, I can see no reason why anyone would explicitly provide an empty string to `delegate_to`, so the workaround is simply to not do that.

Hello, I understand you want to have a consistent behavior for `delegate_to`, and I agree `delegate_to: ""` should not happen.

Our point is more that when the task is skipped because a  `when : my_variable|bool` clause returns false, we would expect ` delegate_to: "{{ old_server }}"` to not fail if the variable is empty. 
It then would be consistent with `become:` keyword. Currently behavior is inconsistent at least between these two keywords `become` and `delegate_to`.

**Comment by s-hertel** [comment]
> Currently behavior is inconsistent at least between these two keywords become and delegate_to.

It's not inherently a bug if two keywords are templated/validated at different points ([related documentation issue](https://github.com/ansible/ansible-documentation/issues/1295)), but seems like it could be possible like this https://github.com/ansible/ansible/commit/c2d5a489b5aa7b1e7c7117b599aac28a6367be46. The block of code between delegate_to validation and evaluating when handles the PlayContext variables, which is mainly for old plugins that haven't been updated to use modern `get_option` syntax, so my only concern would be backwards compatibility.

**Comment by s-hertel** [comment]
I do feel like it would be odd to intentionally support invalid values conditionally. Setting old_server to "{{ inventory_hostname }}" or null to mean no delegation, rather than "" (which looks like an error) would be correct.

**Comment by bcoca** [comment]
or `omit`

**Comment by craph** [comment]
The question is : why `delegate_to` is evaluated if the conditional `when` is evaluated to `false` ?

I was thinking that the when conditional is used to drive the execution or not of some tasks. 

If there is `when: my_variable|bool` and my_variable evaluated to false all the surrounded task or block should be skipped.

Am I understanding it the wrong way ?

Thank you very much for your help.

Best regards,

**Comment by s-hertel** [comment]
> The question is : why delegate_to is evaluated if the conditional when is evaluated to false ?

Because templating `delegate_to` occurs before `when` is evaluated, and the delegate variables can change the result of `when`.

**Comment by bcoca** [comment]
several things in a task get templated before we can evaluate the `when`, those have to be correctly formed even if that task is skipped.

For example,  `when: ansible_delegated_vars['is_primary']`

If we disable the templating in delegation and wait until after the conditional, the above clause will always be invalid and there is no way to make task execution conditional to the delegated host.

**Comment by mat-m** [comment]
@s-hertel @bcoca : Can you confirm my understanding below ?

I should set `delegate_to: "{{ old_server | default(omit) }}". 

When `do_migration` is false and we don't provide an `old_server` value, the `delegate_to` attribute will be ignored, which has no impacts since we will not run the task.

**Comment by s-hertel** [comment]
@mat-m If your `old_server` variable is defined (such as an empty string ""), you'd need to pass the second argument to the `default` filter to override an empty var: `delegate_to: "{{ old_server | default(null, True) }}"`. You might be able to use `inventory_hostname`/`null`/`omit` interchangeably, for simplicity I'd probably use either of the first two options or validate `omit` works for your scenario.

**Comment by mat-m** [comment]
With just `delegate_to: "{{ old_server }}"`:

If we do not define `old_server: ""` in our defaults (losing track of its usage), it works fine as is (without any default).

If we define `old_server: null`, that did not work but raised the following:
`fatal: [localhost -> {{ old_server }}]: FAILED! => {"msg": "Empty hostname produced from delegate_to: \"{{ old_server }}\""}`
I would expect that to have the same behavior as `default(null,True)`, but apparently not, although I don't know why.

**Comment by craph** [comment]
@s-hertel after some testing too, I have the same behavior as @mat-m .
If I remove the declaration of `my_var` (losing track of its usage) it works without having any default defined. 
Also, I tried to define `my_var: null` but it seems it doesn't work with the same behavior as `default(null, True)`. 

Is it normal ?
Thank you very much for your help.

**Comment by craph** [comment]
@s-hertel any  idea about my previous comment ?
